### ISSUE TITLE
Build error on Android NDK r10d

### ISSUE THREAD

**Issue by NuLL3rr0r** [NoAvailableAlias/nano-signal-slot#13]
Upon including nano_signal_slot.hpp in my Cocos2d-x project on Android I get (both Clang 3.6 and GCC 4.9):

In file included from jni/../../Classes/Android.hpp:11:0,
                 from jni/../../Classes/AppDelegate.cpp:4:
jni/../../Classes/nano_signal_slot.hpp:71:39: error: declaration of 'class ... A
rgs'
     template <typename T, typename F, typename ...Args>
                                       ^
jni/../../Classes/nano_signal_slot.hpp:11:24: error:  shadows template parm 'cla
ss ... Args'
 template <typename RT, typename... Args>
                        ^
jni/../../Classes/nano_signal_slot.hpp: In member function 'void Nano::Signal<RT
(Args ...)>::connect(T&)':
jni/../../Classes/nano_signal_slot.hpp:74:20: error: 'mem_ptr' was not declared
in this scope
         connect<T, mem_ptr>(std::addressof(instance));
                    ^
In file included from jni/../../Classes/Android.hpp:11:0,
                 from jni/../../Classes/MainMenuScene.cpp:3:
jni/../../Classes/nano_signal_slot.hpp:71:39: error: declaration of 'class ... A
rgs'
     template <typename T, typename F, typename ...Args>
                                       ^
jni/../../Classes/nano_signal_slot.hpp:11:24: error:  shadows template parm 'cla
ss ... Args'
 template <typename RT, typename... Args>
                        ^
jni/../../Classes/nano_signal_slot.hpp: In member function 'void Nano::Signal<RT
(Args ...)>::connect(T&)':
jni/../../Classes/nano_signal_slot.hpp:74:20: error: 'mem_ptr' was not declared
in this scope
         connect<T, mem_ptr>(std::addressof(instance));
                    ^
[armeabi] StaticLibrary  : libcocos3d.a
In file included from jni/../../Classes/Android.hpp:11:0,
                 from jni/../../Classes/Android.cpp:8:
jni/../../Classes/nano_signal_slot.hpp:71:39: error: declaration of 'class ... A
rgs'
     template <typename T, typename F, typename ...Args>
                                       ^
jni/../../Classes/nano_signal_slot.hpp:11:24: error:  shadows template parm 'cla
ss ... Args'
 template <typename RT, typename... Args>
                        ^
jni/../../Classes/nano_signal_slot.hpp: In member function 'void Nano::Signal<RT
(Args ...)>::connect(T&)':
jni/../../Classes/nano_signal_slot.hpp:74:20: error: 'mem_ptr' was not declared
in this scope
         connect<T, mem_ptr>(std::addressof(instance));
                    ^

If I build for Windows using VS2013 SP5 there are no issues at all.

**Comment by NoAvailableAlias** [comment]
Hmm this is pretty odd, haven't run into this with my gcc testing.
Tested without any issues when compiled under c++11 and c++14.
[http://coliru.stacked-crooked.com/a/f9a36e4049eb1825](coliru gcc 5.2)
[https://ideone.com/eRRQ26](ideone gcc v?)
Will continue to see what I can find.

**Comment by NuLL3rr0r** [comment]
Yes, that's one of the weirdest things I've ever seen.
Since I do development using Qt 5.5.1 for Android to I decided to test it on Qt / Qt Creator too (I'm using this package qt-opensource-windows-x86-android-5.5.1.exe which is shipped along with MinGW 4.9.2).

I created an empty Qt Quick Application project and inside man.cpp I only included nano_signal_slot.hpp without writing any other line of code. Now, it doesn't matter if I build for Android or Desktop it is the same. This is the build output for Desktop(Windows):

01:29:05: Running steps for project test-nano-qt...
01:29:05: Configuration unchanged, skipping qmake step.
01:29:05: Starting: "E:\dev\_tools\Qt5.5.1\Tools\mingw492_32\bin\mingw32-make.exe" 
E:\dev\_tools\Qt5.5.1\5.5\mingw492_32\bin\qmake.exe -spec win32-g++ "CONFIG+=debug" "CONFIG+=declarative_debug" "CONFIG+=qml_debug" -o Makefile ..\test-nano-qt\test-nano-qt.pro
E:/dev/_tools/Qt5.5.1/Tools/mingw492_32/bin/mingw32-make -f Makefile.Debug
mingw32-make[1]: Entering directory 'E:/dev/build-test-nano-qt-Desktop_Qt_5_5_1_MinGW_32bit-Debug'
g++ -c -pipe -fno-keep-inline-dllexport -g -std=c++0x -frtti -Wall -Wextra -fexceptions -mthreads -DUNICODE -DQT_QML_DEBUG -DQT_DECLARATIVE_DEBUG -DQT_QUICK_LIB -DQT_GUI_LIB -DQT_QML_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB -DQT_NEEDS_QMAIN -I..\test-nano-qt -I. -I..\_tools\Qt5.5.1\5.5\mingw492_32\include -I..\_tools\Qt5.5.1\5.5\mingw492_32\include\QtQuick -I..\_tools\Qt5.5.1\5.5\mingw492_32\include\QtGui -I..\_tools\Qt5.5.1\5.5\mingw492_32\include\QtANGLE -I..\_tools\Qt5.5.1\5.5\mingw492_32\include\QtQml -I..\_tools\Qt5.5.1\5.5\mingw492_32\include\QtNetwork -I..\_tools\Qt5.5.1\5.5\mingw492_32\include\QtCore -Idebug -I..\_tools\Qt5.5.1\5.5\mingw492_32\mkspecs\win32-g++  -o debug\main.o ..\test-nano-qt\main.cpp
In file included from ..\test-nano-qt\main.cpp:3:0:
..\test-nano-qt\nano_signal_slot.hpp:71:39: error: declaration of 'class ... Args'
     template <typename T, typename F, typename ...Args>
                                       ^
..\test-nano-qt\nano_signal_slot.hpp:11:24: error:  shadows template parm 'class ... Args'
 template <typename RT, typename... Args>
                        ^
..\test-nano-qt\nano_signal_slot.hpp:144:19: error: expected ')' before '&&' token
     void emit(Uref&&... args)
                   ^
..\test-nano-qt\nano_signal_slot.hpp: In member function 'void Nano::Signal<RT(Args ...)>::connect(T&)':
..\test-nano-qt\nano_signal_slot.hpp:74:20: error: 'mem_ptr' was not declared in this scope
         connect<T, mem_ptr>(std::addressof(instance));
                    ^
Makefile.Debug:217: recipe for target 'debug/main.o' failed
mingw32-make[1]: *** [debug/main.o] Error 1
mingw32-make[1]: Leaving directory 'E:/dev/build-test-nano-qt-Desktop_Qt_5_5_1_MinGW_32bit-Debug'
mingw32-make: *** [debug] Error 2
makefile:34: recipe for target 'debug' failed
01:29:08: The process "E:\dev\_tools\Qt5.5.1\Tools\mingw492_32\bin\mingw32-make.exe" exited with code 2.
Error while building/deploying project test-nano-qt (kit: Desktop Qt 5.5.1 MinGW 32bit)
When executing step "Make"
01:29:08: Elapsed time: 00:03.

**Comment by NuLL3rr0r** [comment]
I invested a little bit more time into this. At first I thought there might be a bug or something in NDK or MinGW, so, I ran another test on my FreeBSD system which uses Clang as its default compiler using the following code:


#include <QGuiApplication>
#include <QQmlApplicationEngine>
#include "nano_signal_slot.hpp"

int main(int argc, char *argv[])
{
    QGuiApplication app(argc, argv);

    QQmlApplicationEngine engine;
    engine.load(QUrl(QStringLiteral("qrc:/main.qml")));

    return app.exec();
}


02:49:57: Running steps for project test-nano...
02:49:57: Starting: "/usr/local/lib/qt5/bin/qmake" /usr/home/babaei/Desktop/test-nano/test-nano.pro -r -spec freebsd-clang CONFIG+=debug CONFIG+=declarative_debug CONFIG+=qml_debug
02:49:57: The process "/usr/local/lib/qt5/bin/qmake" exited normally.
02:49:57: Starting: "/usr/bin/make" 
clang++ -c -pipe -g -std=c++11 -Wall -W -pthread -D_THREAD_SAFE -fPIE -DQT_QML_DEBUG -DQT_DECLARATIVE_DEBUG -DQT_QUICK_LIB -DQT_GUI_LIB -DQT_QML_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB -I../test-nano -I. -I../../../../local/include/qt5 -I../../../../local/include/qt5/QtQuick -I../../../../local/include/qt5/QtGui -I../../../../local/include/qt5/QtQml -I../../../../local/include/qt5/QtNetwork -I../../../../local/include/qt5/QtCore -I. -I../../../../local/lib/qt5/mkspecs/freebsd-clang -o main.o ../test-nano/main.cpp
In file included from ../test-nano/main.cpp:3:
In file included from ../test-nano/nano_signal_slot.hpp:4:
../test-nano/nano_function.hpp:58:18: warning: suggest braces around initialization of subobject [-Wmissing-braces]
        return { reinterpret_cast<std::uintptr_t>(m_this_ptr),
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../test-nano/main.cpp:3:
../test-nano/nano_signal_slot.hpp:71:51: error: declaration of 'Args' shadows template parameter
    template <typename T, typename F, typename ...Args>
                                                  ^
../test-nano/nano_signal_slot.hpp:11:36: note: template parameter is declared here
template <typename RT, typename... Args>
                                   ^
../test-nano/nano_signal_slot.hpp:144:19: error: expected ')'
    void emit(Uref&&... args)
                  ^
../test-nano/nano_signal_slot.hpp:144:14: note: to match this '('
    void emit(Uref&&... args)
             ^
../test-nano/nano_signal_slot.hpp:144:15: error: member 'Uref' declared as a template
    void emit(Uref&&... args)
              ^
../test-nano/nano_signal_slot.hpp:74:20: error: use of undeclared identifier 'mem_ptr'
        connect<T, mem_ptr>(std::addressof(instance));
                   ^
../test-nano/nano_signal_slot.hpp:72:21: warning: unused parameter 'instance' [-Wunused-parameter]
    void connect(T& instance)
                    ^
../test-nano/nano_signal_slot.hpp:147:6: error: expected ';' after class
    }
     ^
     ;
*** Error code 1

Stop.
make: stopped in /usr/home/babaei/Desktop/build-test-nano-Desktop-Debug
../test-nano/nano_signal_slot.hpp:152:37: error: use of undeclared identifier 'Delegate'
        Observer::onEach_Accumulate<Delegate, Accumulate>
                                    ^
../test-nano/nano_signal_slot.hpp:153:39: error: unknown type name 'accumulate'; did you mean 'Accumulate'?
            (std::forward<Accumulate>(accumulate), std::forward<Uref>(args)...);
                                      ^~~~~~~~~~
                                      Accumulate
../test-nano/nano_signal_slot.hpp:149:24: note: 'Accumulate' declared here
    template <typename Accumulate, typename... Uref>
                       ^
../test-nano/nano_signal_slot.hpp:153:50: error: expected ')'
            (std::forward<Accumulate>(accumulate), std::forward<Uref>(args)...);
                                                 ^
../test-nano/nano_signal_slot.hpp:153:13: note: to match this '('
            (std::forward<Accumulate>(accumulate), std::forward<Uref>(args)...);
            ^
../test-nano/nano_signal_slot.hpp:153:19: error: definition or redeclaration of 'forward' not allowed inside a function
            (std::forward<Accumulate>(accumulate), std::forward<Uref>(args)...);
             ~~~~~^
../test-nano/nano_signal_slot.hpp:153:19: error: C++ requires a type specifier for all declarations
            (std::forward<Accumulate>(accumulate), std::forward<Uref>(args)...);
            ~     ^
../test-nano/nano_signal_slot.hpp:150:39: warning: unused parameter 'accumulate' [-Wunused-parameter]
    void emit_accumulate(Accumulate&& accumulate, Uref&&... args)
                                      ^
../test-nano/nano_signal_slot.hpp:150:61: warning: unused parameter 'args' [-Wunused-parameter]
    void emit_accumulate(Accumulate&& accumulate, Uref&&... args)
                                                            ^
../test-nano/nano_signal_slot.hpp:160:26: error: call to non-static member function without an object argument
        return Observer::empty();
               ~~~~~~~~~~^~~~~
../test-nano/nano_signal_slot.hpp:160:26: error: 'empty' is a private member of 'Nano::Observer'
../test-nano/nano_observer.hpp:54:10: note: implicitly declared private here
    bool empty()
         ^
In file included from ../test-nano/main.cpp:3:
../test-nano/nano_signal_slot.hpp:165:1: error: extraneous closing brace ('}')
} // namespace Nano ------------------------------------------------------------
^
4 warnings and 13 errors generated.
02:49:57: The process "/usr/bin/make" exited with code 1.
Error while building/deploying project test-nano (kit: Desktop)
When executing step "Make"
02:49:57: Elapsed time: 00:01.


I've got an Slackware system too, which I do not have access to right now. It uses GCC. If the result is going to be different I'll post it here.

**Comment by NoAvailableAlias** [comment]
I'm not going to lie I'm not finding much time to work on work I want to work on outside of work, (not to mention my lack of build environment diversity), however, it looks like whatever environment you are developing on is not liking the non-type template parameters.

If it is possible, could you try compiling a "minimal" stripped down nano-signal-slot [testcase found here.](https://ideone.com/sPFLH4)

Could you also try compiling this [small non-type template parameter snippet](http://ideone.com/PE2VrK)? ([also found here](http://en.cppreference.com/w/cpp/language/template_parameters))

**Comment by NoAvailableAlias** [comment]
Well there's got to be something not entirely standard for this to be happening. Will start by fixing that brace initializer warning and shadowing.

**Comment by NuLL3rr0r** [comment]
Are you suggesting that's a bug in the compilers (except Visual Studio of course)?
If so, it's not a bad idea to report them.

**Comment by NoAvailableAlias** [comment]
No, that usually isn't the case. It's usually something your code is doing that may be slightly non standard where a particular compiler doesn't like it. But I'm not seeing where exactly. Again it doesn't help I can't reproduce with my lack of build environments.

**Comment by MarkoTurunen** [comment]
I am using nano-signal-slot with cocos2d-x and Android NDK r10d without issues. Maybe something wrong in your build files?

**Comment by NoAvailableAlias** [comment]
This issue is most likely no longer relevant even for nano-signal-slot v1.x due to compiler changes.

This issue will be closed once v2.0 milestone is complete.
